src/broadcast/broadcast.rs

```diff
@@ -284,21 +384,129 @@ impl<N: NodeIdT> Broadcast<N> {
             // Enqueue a broadcast of a Ready message.
             step.extend(self.send_ready(hash)?);
         }
+        // Upon receiving 2f + 1 matching Ready(h) messages, send full
+        // `Echo` message to every node who hasn't sent us a `CanDecode`
+        if self.count_readys(hash) == 2 * self.netinfo.num_faulty() + 1 {
+            step.extend(self.send_echo_remaining(hash)?);
+        }
+
         Ok(step.join(self.compute_output(hash)?))
     }
 
-    /// Sends an `Echo` message and handles it. Does nothing if we are only an observer.
-    fn send_echo(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
-        self.echo_sent = true;
+    /// Sends `Echo` message to all left nodes and handles it.
+    fn send_echo_left(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
         if !self.netinfo.is_validator() {
             return Ok(Step::default());
         }
         let echo_msg = Message::Echo(p.clone());
-        let step: Step<_> = Target::All.message(echo_msg).into();
+        let mut step = Step::default();
+        // `N - 2f + g` node ids to the left of our_id (excluding our_id)
+        // after arranging all node ids in a circular list.
+        let left = self
+            .netinfo
+            .all_ids()
+            .cycle()
+            .skip_while(|x| *x != self.our_id())
```

Comment:
`.skip(self.netinfo.node_index(self.our_id()))` should be faster and equivalent? (Also below.)

Reply:
`node_index` returns an `Option<usize>`. This way seems easier rather than unwrapping.